manager/src/main/java/org/openremote/manager/rules/AssetQueryPredicate.java

```diff
@@ -184,65 +185,114 @@ public static Predicate<NameValueHolder<?>> asPredicate(Supplier<Long> currentMi
      * {@link LogicGroup.Operator#AND} to be applied.
      * @return The matched asset states or null if there is no match
      */
-    @SuppressWarnings("unchecked")
     public static Function<Collection<AttributeInfo>, Set<AttributeInfo>> asAttributeMatcher(Supplier<Long> currentMillisProducer, LogicGroup<AttributePredicate> condition) {
+        return asAttributeMatcher(currentMillisProducer, condition, null);
+    }
+
+    /**
+     * A function for matching {@link AttributeInfo}s of an asset; the infos must be related to the same asset to allow
+     * {@link LogicGroup.Operator#AND} to be applied.
+     * {@link AttributeConditionTimeState} can be provided to restrict the match to a certain time duration based
+     * @return The matched asset states or null if there is no match
+     */
+    @SuppressWarnings("unchecked")
+    public static Function<Collection<AttributeInfo>, Set<AttributeInfo>> asAttributeMatcher(
+            Supplier<Long> currentMillisProducer, 
+            LogicGroup<AttributePredicate> condition,
+            AttributeConditionTimeState attributeConditionTimeState) {
+
         if (groupIsEmpty(condition)) {
-            return as -> Collections.EMPTY_SET;
+            return as -> Collections.emptySet();
         }
 
         LogicGroup.Operator operator = condition.operator == null ? LogicGroup.Operator.AND : condition.operator;
         List<Function<Collection<AttributeInfo>, Set<AttributeInfo>>> assetStateMatchers = new ArrayList<>();
-        List<Predicate<AttributeInfo>> attributePredicates = new ArrayList<>();
+        Map<Integer, List<Predicate<AttributeInfo>>> attributePredicates = new HashMap<>();
 
-        if (condition.getItems().size() > 0) {
+        if (!condition.getItems().isEmpty()) {
+            condition.getItems().stream().forEach(p -> {
+                int index = condition.getItems().indexOf(p);
+                attributePredicates.put(index, new ArrayList<>());
+                attributePredicates.get(index).add((Predicate<AttributeInfo>) (Predicate) asPredicate(currentMillisProducer, p));
```

Comment:
This new map and attributeConditionTimeState object seems un-neccessary:

Couldn't you do this:

```
attributePredicates.add((Predicate<AttributeInfo>)(Predicate)asPredicate(currentMillisProducer, p, ruleCondition.duration != null ? ruleCondition.duration.get(i) : null));
```

All state can just be fields within the predicate rather than having the `AttributeConditionTimeState` `Map`

Reply:
Removed both the map and TimeState

Duration is directly passed to the asPredicate, and the predicate now deals with the duration match state.